airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java

```diff
@@ -361,11 +367,27 @@ public WebBackendConnectionRead webBackendUpdateConnection(final WebBackendConne
     connectionRead = connectionsHandler.updateConnection(connectionUpdate);
 
     if (needReset) {
+      final UUID connectionId = webBackendConnectionUpdate.getConnectionId();
+      final ConnectionIdRequestBody connectionIdRequestBody = new ConnectionIdRequestBody().connectionId(connectionId);
+      final ConnectionStateType stateType = getStateType(connectionIdRequestBody);
+      final List<io.airbyte.protocol.models.StreamDescriptor> streamsToReset;
+
+      if (stateType == ConnectionStateType.LEGACY || stateType == ConnectionStateType.NOT_SET) {
+        streamsToReset = configRepository.getAllStreamsForConnection(connectionId);
+      } else {
+        final ConfiguredAirbyteCatalog existingConfiguredCatalog =
+            configRepository.getConfiguredCatalogForConnection(connectionId);
+        final io.airbyte.protocol.models.AirbyteCatalog existingCatalog = CatalogHelpers.configuredCatalogToCatalog(existingConfiguredCatalog);
+        final AirbyteCatalog apiExistingCatalog = CatalogConverter.toApi(existingCatalog);
+        final AirbyteCatalog newAirbyteCatalog = webBackendConnectionUpdate.getSyncCatalog();
+        final CatalogDiff catalogDiff = connectionsHandler.getDiff(apiExistingCatalog, newAirbyteCatalog);
+        final List<StreamDescriptor> apiStreamsToReset = getStreamsToReset(catalogDiff);
+        streamsToReset = apiStreamsToReset.stream().map(streamDescriptor -> ProtocolConverters.streamDescriptorToProtocol(streamDescriptor)).toList();
+      }
+
       ManualOperationResult manualOperationResult = eventRunner.synchronousResetConnection(
           webBackendConnectionUpdate.getConnectionId(),
-          // TODO (https://github.com/airbytehq/airbyte/issues/12741): change this to only get new/updated
-          // streams, instead of all
-          configRepository.getAllStreamsForConnection(webBackendConnectionUpdate.getConnectionId()));
+          streamsToReset);
```

Comment:
We should probably check here if `streamsToReset` is empty, and if so do not call `synchronousResetConnection` or `startNewManualSync`. Otherwise, in the case that there is nothing to reset, then we will just cancel any currently running job and start a new normal sync for no reason. 

We should also add a test case for this

Reply:
I thought we sort of had that check with the `needsReset` bool above?

Reply:
That `needsReset` boolean is just checking if the `withRefreshedCatalog` flag is set on the input to this endpoint, which I believe the webapp sets whenever the `Refresh source schema` button is pressed in the UI and the user saves the connection settings. So I believe it is possible that the user refreshes the schema, but nothing in the schema has actually changed, and then saves the connection, causing this `withRefreshedCatalog` flag being set to true despite nothing changing.

Reply:
ah is that how it works? so currently, if there is a schema refresh but no actual change, we reset all streams?

Reply:
I did update this, but still curious how this isn't resetting all streams on refresh right now

Reply:
Yeah we probably are resetting all streams right now if that situation happens. Not 100% sure though, could be an interesting thing to test